File: go/lib/bron_kerbosch3_gp.go
package BronKerbosch

func bronKerbosch3gp(graph *UndirectedGraph, reporter Reporter) {
	// Bron-Kerbosch algorithm with degeneracy ordering, with nested searches
	// choosing a pivot from candidates only (IK_GP).
	var ordering SimpleVertexVisitor
	degeneracyOrdering(graph, &ordering, -1)
	excluded := make(VertexSet, len(ordering.vertices))
	for _, v := range ordering.vertices {
		neighbours := graph.neighbours(v)
		neighbouringCandidates := neighbours.Difference(excluded)
		if !neighbouringCandidates.IsEmpty() {
			neighbouringExcluded := neighbours.Intersection(excluded)
			visit(
				graph, reporter,
				MaxDegree, MaxDegreeLocal,
				neighbouringCandidates,
				neighbouringExcluded,
				[]Vertex{v})
		}
		excluded.Add(v)
	}
}

File: rust/bron_kerbosch/src/core/bron_kerbosch3_gp.rs
//! Bron-Kerbosch algorithm with degeneracy ordering, with nested searches
//! choosing a pivot from candidates only (IK_GP)

use super::bron_kerbosch_degen::{PivotChoice, explore_with_degeneracy};
use super::clique_consumer::CliqueConsumer;
use super::graph::Graph;
use super::vertexsetlike::VertexSetLike;

pub fn explore<VertexSet, Consumer>(
    graph: &Graph<VertexSet>,
    consumer: Consumer,
) -> Consumer::Harvest
where
    VertexSet: VertexSetLike,
    Consumer: CliqueConsumer,
{
    explore_with_degeneracy(graph, consumer, PivotChoice::MaxDegreeLocal)
}
